sw/device/lib/dif/dif_template.h

```diff
@@ -0,0 +1,253 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_TEMPLATE_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_TEMPLATE_H_
+
+// This file is the "DIF Library header template", which provides a base for
+// building a DIF for a new peripheral, defining all of the declatations that
+// would be expected of a DIF library in the README.md.
+//
+// To instatntiate this for a new IP named my_ip, simply:
+// - Copy this file into dif_my_ip.h.
+// - Replace all occurences of `<ip>` with `my_ip`.
+// - Replace all occurences of `<IP>` with `MyIp`.
+// - Replace all occurences of `<Peripheral>` and `<peripheral>` with a
+//   documentation-appropriate name for my_ip, like "my peripheral"; mind the
+//   capitalization. Proof-reading the comments may be necessary.
+// - Replace the `handle` function parameter, if you want.
+// - Delete any definitions that are not required for your peripheral.
+// - Delete this comment and the guard #error below.
+//
+// Note that this file is unlikely to parse as real C at all.
+
+#error "This file is a template, and not real code."
+
+/**
+ * @file
+ * @brief <a href="/hw/ip/i2c/doc/"><Peripheral></a> Device Interface Functions
+ */
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * A toggle state: enabled, or disabled.
+ *
+ * This enum may be used instead of a `bool` when describing an enabled/disabled
+ * state.
+ */
+typedef enum dif_<ip>_toggle {
+  /**
+   * The "enabled" state.
+   */
+  kDifI2cToggleEnabled,
+  /**
+   * The "disabled" state.
+   */
+  kDifI2cToggleDisabled,
+} dif_<i2c>_toggle_t;
+
+/**
+ * Hardware instantiation parameters for <peripheral>.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of the a toplevel.
+ */
+typedef struct dif_<ip>_param {
+  /**
+   * The base address for the <peripheral> hardware registers.
+   */
+  mmio_region_t base_addr;
+
+  // Other fields, if necessary.
+} dif_<ip>_param_t;
+
+/**
+ * Runtime configuration for <peripheral>.
+ *
+ * This struct describes runtime information for one-time configuration of the
+ * hardware.
+ */
+typedef struct dif_<ip>_config {
+  /**
+   * The base address for the <peripheral> hardware registers.
+   */
+  mmio_region_t base_addr;
+
+  // Other fields, if necessary.
+} dif_<ip>_config_t;
+
+/**
+ * A handle to <peripheral>.
+ *
+ * This type should be treated as opaque by users.
+ */
+typedef struct dif_<ip> {
+  dif_param<ip> params;
```

Comment:
Correction: use the typedef here
```suggestion
  dif_param_<ip>_t params;
```